customer-service/src/main/java/io/eventuate/examples/tram/ordersandcustomers/customers/service/CustomerService.java

```diff
@@ -0,0 +1,115 @@
+package io.eventuate.examples.tram.ordersandcustomers.customers.service;
+
+import io.eventuate.examples.tram.ordersandcustomers.common.domain.Money;
+import io.eventuate.examples.tram.ordersandcustomers.customers.domain.Customer;
+import io.eventuate.examples.tram.ordersandcustomers.customers.domain.CustomerRepository;
+import io.eventuate.examples.tram.ordersandcustomers.customers.domain.ReservedCredit;
+import io.eventuate.examples.tram.ordersandcustomers.customers.domain.ReservedCreditRepository;
+import io.eventuate.examples.tram.ordersandcustomers.customers.domain.events.CustomerCreditReservationFailedEvent;
+import io.eventuate.examples.tram.ordersandcustomers.customers.domain.events.CustomerCreditReservedEvent;
+import io.eventuate.examples.tram.ordersandcustomers.customers.domain.events.CustomerValidationFailedEvent;
+import io.eventuate.tram.events.publisher.ResultWithEvents;
+import io.eventuate.tram.messaging.common.Message;
+import io.eventuate.tram.spring.events.publisher.ReactiveDomainEventPublisher;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.transaction.reactive.TransactionalOperator;
+import reactor.core.publisher.Mono;
+
+import java.math.BigDecimal;
+import java.util.Collections;
+import java.util.List;
+
+public class CustomerService {
+
+  private Logger logger = LoggerFactory.getLogger(getClass());
+
+  private CustomerRepository customerRepository;
+
+  private ReservedCreditRepository reservedCreditRepository;
+
+  private ReactiveDomainEventPublisher domainEventPublisher;
+
+  private TransactionalOperator transactionalOperator;
+
+  public CustomerService(CustomerRepository customerRepository,
+                         ReservedCreditRepository reservedCreditRepository,
+                         ReactiveDomainEventPublisher domainEventPublisher,
+                         TransactionalOperator transactionalOperator) {
+    this.customerRepository = customerRepository;
+    this.reservedCreditRepository = reservedCreditRepository;
+    this.domainEventPublisher = domainEventPublisher;
+    this.transactionalOperator = transactionalOperator;
+  }
+
+  public Mono<Customer> createCustomer(String name, Money creditLimit) {
+    ResultWithEvents<Customer> customerWithEvents = Customer.create(name, creditLimit);
+
+    return customerRepository
+            .save(customerWithEvents.result)
+            .flatMap(customer -> domainEventPublisher.publish(Customer.class, customer.getId(), customerWithEvents.events).collectList().map(messages -> customer))
+            .as(transactionalOperator::transactional);
+  }
+
+  public Mono<Void> reserveCredit(long orderId, long customerId, Money orderTotal) {
+
+    Mono<Customer> possibleCustomer = customerRepository.findById(customerId);
```

Comment:
This doesn't handle the scenario where there are two simultaneously attempts to reserve credit.
Both will read existing and potentially exceed the available credit.

This might be a bug in the existing customer service BTW. The solution would be to implement optimistic locking in the Customer.

Reply:
As I know, optimistic locking works only with single row, but we are reserving credit using separate table. We can use Serializable isolation level. Or probably modify customer to hold a sum of reservations and if insert row to credit_reservation and update the sum in the same transaction, in case if there is concurrent modification, insert will be rolled back together with customer changes. In theory. I need to check docs to recall details about optimistic locking and write a test. Also there could be other useful lock types, for example postgres has Advisory Locks (https://www.postgresql.org/docs/9.1/explicit-locking.html), it could be something similar for mysql, but it definitely would complicate the code.

Reply:
> implement optimistic locking in the Customer

The Customer has a version column, which is incremented whenever a CreditReservation is inserted/deleted.
e.g.

1. Load Customer to get current version
2. Insert/delete Credit Reservation
3. `update Customer set version = version + 1 where id = ? and version = <old version>`

Reply:
and if this update will return 0 updated rows, then rollback, right?

Reply:
Yes. Rollback and retry.

Reply:
Got it, thank you